### PR TITLE
Assign to test failure GH Action

### PR DESCRIPTION
**Pull request by Patras3**
The main assumption is to assign a developer who lastly edited the code mentioned in the test failure issue.

This workflow will be invoked for issues with `.*Test-Failure.*` label. Creating a new issue with this label or adding label later has the same effect.

Implemented steps:
1. Find commit from issue description: `(commit\s([^)\s]+)` with fallback to `master`. 
_This regex meets currently used template by QEs team._
2. Checkout to specific commit or master branch
3. Parse issue title to find affected code
_Examples_:

| Title | Path | 
| --- | --- |
| `KinesisIntegrationTest` | `KinesisIntegrationTest.java` | 
| `KinesisIntegrationTest.dynamicStream_1Shard_splitsDuringData` | `KinesisIntegrationTest.java` | 
| `Several tests failed with NPE in 'com.hazelcast.cache.impl.hidensity.operation'` |`com/hazelcast/cache/impl/hidensity/operation` |
| `Nothing about package and no class name with Test suffix` | -|

4. Find last commit for path find in previous step.
_It uses bash `find . -wholename "*$FILENAME*" -type f` and iterates of each file to find recently edit file (committed)_
5. Find GH username for commit
6. Check if username belongs to organization
7. Print details
8. Evaluate suggested assignment
_Found username from previous steps will be used only if this username belongs currently to the organization. If any of previous step is failed or there is no reference to code in the title, or username is outside of organization, then author of issue will be assigned as fallback._
9. Assign issue and and proper comment
_Types of comments:_
_`Assigned @${commitAuthor} due to being the author of last commit (${commit})`
`Assigned @${issueAuthor}, because identified author (${commitAuthor}) of last commit (${commit}) is not a member of the organization`
`Assigned @${issueAuthor} due to not found reference to code in the issue `_

Checklist:
- [X] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [X] Label `Add to Release Notes` or `Not Release Notes content` set
- [X] Request reviewers if possible

### CODE REVIEW

**Comment by Patras3** [comment]
I would like to merge this, test with some TFs, and temporarily disable it until this way of handling test failures will be officially introduced. In this way, my upcoming 2 weeks' absence won't have an impact.

**Comment by frant-hartm** [comment]
@ldziedziul could you please have a look at this?

**Review by kwart — APPROVED** [review]

**Review by ldziedziul — APPROVED** [review]
In general looks good but I'm concerned about maintainability of these JS scripts embedded in the workflow file. My personal preference would be to use bash scripts instead ;)

File: .github/workflows/assign-to-test-failure.yml [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+            function extractNameFromPackageString(testRef) {
+              //Check if contains class name
+              if (testRef.toLowerCase() !== testRef) {
+                const testPath = testRef.split(".").join("/")
+                //Remove testName/case from name
+                //Regex: Everything before and capitalized word
+                const match = testRef.match(/(.*?[A-Z]\w*)/)
+                return replaceQuotes(match ? match[1] : testRef) + ".java"
+              }
+            //Return package path
```
Comment (olukas) [inline .github/workflows/assign-to-test-failure.yml:85]: nit: alignment

File: .github/workflows/assign-to-test-failure.yml [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+            if ("${{ fromJson(steps.summary.outputs.result).test.lastCommit.author.belongsToOrganization }}" !== "true") {
+              return { 
+                comment: `Assigned @${issueAuthor}, because identified author (${commitAuthor}) of last commit (${commit}) is not a member of the organization`, 
+                finalAssignement: issueAuthor
+              }
+            }
+            //Proper case
+            return {
+              comment: `Assigned @${commitAuthor} due to being the author of last commit (${commit})`, 
+              finalAssignement: issueAuthor
```
Comment (olukas) [inline .github/workflows/assign-to-test-failure.yml:248]: Shouldn't this be `commitAuthor`?

**Review by olukas — CHANGES_REQUESTED** [review]
IIUC there's one copy-paste error which actually means it will work incorrectly for common case - when we find author of commit we will not assign him but still assign `issueAuthor`.

**Comment by Patras3** [comment]
run-ee-compile

**Review by olukas — APPROVED** [review]

**Review by AyberkSorgun — APPROVED** [review]
LGTM, great work!

**Pull request merged**